### DISCUSSION TITLE
splitAtom+focusAtom still re-rendering other elements in array

### DISCUSSION THREAD

**Discussion by utilitydelta** [pmndrs/jotai#1674] [Q&A]
Hi, 
I have an array of tasks which I use splitAtom on. Then I use focusAtom to focus on 'textdata' - a property on the task. It is then bound to an input element with value and onChange. 
Performance when typing in the <input> element is very slow. Seems to be re-rendering every task on each key input?

Click 'Load data!' button and type some text into one of the input elements.
https://codesandbox.io/s/frosty-jasper-5tfzgu?file=/src/App.js

Thanks,
Tyson

**Comment by dai-shi** [comment]
The code looks good. `React.memo` helps a little bit. Maybe, splitAtom is slow?
https://codesandbox.io/s/quizzical-wu-s0qn4k

Can anyone dig more into it?

**Reply by utilitydelta** [reply]
I tried a few different permutations, focusAtom also displays the same behaviour:
https://codesandbox.io/s/quirky-water-kizjvo?file=/src/App.js

This is the only version that performs well:
https://codesandbox.io/s/gracious-frog-xp1yfo?file=/src/App.js:554-586

The takeaway I got from this is utils like splitAtom and focusAtom carry unexpected dangers; they do not behave as independently updatable atoms. If I want to avoid performance issues with large arrays or objects, I'll need to stick to creating an atom for each element.

**Reply by dai-shi** [reply]
> I tried a few different permutations, focusAtom also displays the same behaviour: https://codesandbox.io/s/quirky-water-kizjvo?file=/src/App.js

Oh, nice finding. So, focusAtom is rather slow, not splitAtom.
https://codesandbox.io/s/agitated-wave-5ih2g7
hm, a little better, but still laggy.

> This is the only version that performs well: https://codesandbox.io/s/gracious-frog-xp1yfo?file=/src/App.js:554-586

Yeah, I would use it (atoms-in-atom) too if performance is concern.

**Reply by dai-shi** [reply]
> https://codesandbox.io/s/agitated-wave-5ih2g7
> hm, a little better, but still laggy.

https://codesandbox.io/s/peaceful-edison-04uc3r
Tried the same one with pre-released v2 API. This one is actually good.
The new `splitAtom` might have fixed some performance issue.

**Reply by utilitydelta** [reply]
splitAtom by itself (without any optics) also is laggy in current release:
https://codesandbox.io/s/condescending-carson-1h2c1p?file=/src/App.js

Thanks for trying out the v2 API with my example. Exciting to see it working as intended! Maybe I'll just upgrade to v2 now then :)

As an aside - this style of state management is actually very similar to ECS in game development - ie. using Structure of Arrays instead of Arrays of Structures (https://en.wikipedia.org/wiki/AoS_and_SoA). Would actually be a cool project to build an ECS on top of jotai / react!

**Reply by dai-shi** [reply]
speaking of game engine, while I'm not familiar with it at all, I created this: https://github.com/jotai-labs/jotai-game
would be nice to have more features for game-like apps.

**Reply by utilitydelta** [reply]
very cool. at a previous company we were using ECS to efficiently render large 3d models for mining engineering CAD software. The main purpose is to align data in memory to make it more cache friendly. There are parallels with react/jotai atoms though, it feels like the same thing - instead of optimising for CPU cache prefetching we would be optimising for react rendering. I'll have a play and see if I can come up with something generic that could be used as a library!
